contracts/dtoken/src/user_flow_protection.rs

```diff
@@ -0,0 +1,103 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::{AccountId, BlockHeight, log, near_bindgen};
+use near_sdk::collections::{UnorderedMap};
+use near_sdk::env::block_height;
+
+const BLOCKS_TO_NEXT_OPERATION: BlockHeight = 100;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct UserFlowProtection {
+    blocked_accounts: UnorderedMap<AccountId, BlockHeight>,
+}
+
+impl Default for UserFlowProtection {
+    fn default() -> Self {
+        Self {
+            blocked_accounts: UnorderedMap::new(b"s".to_vec()),
+        }
+    }
+}
+
+#[near_bindgen]
+impl UserFlowProtection {
+    pub fn block_account(&mut self, account_id: AccountId) {
+        let block_height = block_height();
+        log!("Block account: {} at block index: {}", account_id, block_height);
+        self.blocked_accounts.insert(&account_id, &block_height);
+    }
+
+    pub fn unblock_account(&mut self, account_id: AccountId) {
+        log!("Unblock operation for account: {}", account_id);
+        self.blocked_accounts.remove(&account_id);
+    }
+
+    pub fn get_last_block_index(&self, account_id: AccountId) -> BlockHeight {
+        match self.blocked_accounts.get(&account_id) {
+            Some(index) => index,
+            None => 0
+        }
+    }
+
+    pub fn is_user_can_perform_operation(&mut self, account_id: AccountId) -> bool {
```

Comment:
@TymofiiSavinov we also would need something like `try_lock -> bool` that return `false` if account locked and timeout has not yet expired (aka is_user_can_perform_operation) and `true` otherwise